sympy/matrices/sparse.py

```diff
@@ -158,17 +162,31 @@ def update(i, j, v):
                         smat[i, j] = v
 
                 # manual copy, copy.deepcopy() doesn't work
-                for (r, c), v in args[2].items():
-                    if isinstance(v, MatrixBase):
-                        for (i, j), vv in v.todok().items():
-                            update(r + i, c + j, vv)
-                    elif isinstance(v, (list, tuple)):
-                        _, _, smat = cls._handle_creation_inputs(v, **kwargs)
-                        for i, j in smat:
-                            update(r + i, c + j, smat[i, j])
-                    else:
-                        v = cls._sympify(v)
-                        update(r, c, cls._sympify(v))
+                try:
+                    for (r, c), v in args[2].items():
+                        if isinstance(v, MatrixBase):
+                            for (i, j), vv in v.todok().items():
+                                update(r + i, c + j, vv)
+                        elif isinstance(v, (list, tuple)):
+                            _, _, smat = cls._handle_creation_inputs(v, **kwargs)
+                            for i, j in smat:
+                                update(r + i, c + j, smat[i, j])
+                        else:
+                            v = cls._sympify(v)
+                            update(r, c, cls._sympify(v))
+                except TypeError:  # maybe dod format
```

Comment:
This function is already very complicated and this is making it worse.

Reply:
It would be better to make an explicit `from_dod` function for `Matrix`.

Reply:
I prefer to see this economical representation recognized directly. It's a pretty unambiguous form but our current state of deprecation would allow a dictionary to occupy a Matrix entry. I have removed this code from SparseMatrix, however.

Reply:
These changes have been merged in #23699 though.

Reply:
I'm trying to revert that now. That was unintentional.